### PR TITLE
[WIP][Keybase] Mini-keybase implementation

### LINKED ISSUE #150: [Keybase] Miniature Keybase Implementation to Enable M1 - Create and Delete
# Objective
Create a miniature Keybase that can create, store, and delete keys

# Origin Document
In order to achieve the M1 milestone a mini-keybase implementation is needed to enable dynamic accounts and actors.

This keybase may be used for the following tasks:
- Signing blocks
- Signing transactions
- Signing arbitrary messages
- Signing multi-sig messages

# Goals / Deliverables
Mini ED-25519 Keybase: 
- [ ] Code complete implementation of:
   - [ ] **CREATE** and store keys
   - [ ] **DELETE** keys
- [ ] Document the key architecture 

# Non-Goals / Deliverables
- Integrating Keybase with the CLI or other parts of the codebase
- Implementing multisig

## General issue checklist
- [ ] Update the appropriate CHANGELOG
- [ ] Update the README
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [ ] Update any relevant global documentation & references
- [ ] Document small issues / TODOs along the way

---

Creator: @andrewnguyen22

### PR DESCRIPTION
**Pull request by Jasonyou1995**
## Description

Implementing Mini ED-25519 Keybase:

 Code complete implementation of:
 - [x] CREATE and store keys
 - [x] DELETE keys
 - [ ] Document the key architecture

## Issue

Fixes #150 

## Type of change

Please mark the relevant option(s):

- [x]  New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

## List of changes

- Implementing the mini keybase package
- Create Key
- Store Key
- Delete Key

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by Olshansk** [comment]
@Jasonyou1995 I see that you added myself and Andrew as reviewers. Are you looking for us to review this PR?

**Comment by Jasonyou1995** [comment]
> 3\. Th

Thanks for the suggestions @Olshansk ! I will checkout the branches created by @deblasis locally. 

The codes I'm working on are not yet ready for review. I will switch from draft to formal PR once the codes are ready. I will specify the parts that need to be reviewed.  Thanks for the good examples!

**Comment by Olshansk** [comment]
@Jasonyou1995 If you change the base, it'll help us only see the differences. Once the other PR is merged, we can change it back.

Screen Shot 2022-11-09 at 11 10 26 AM

Also, an idea I had is adding a target to our `Makefile` that can be invoked with `make create_key` that'll be an easy entrypoint for the CLI call with some defaults in place. Not a must, but it'd be an easy way for us to pull this branch down and test out the code locally.

**Comment by Jasonyou1995** [comment]
> @Jasonyou1995 If you change the base, it'll help us only see the differences. Once the other PR is merged, we can change it back.
> 
> Screen Shot 2022-11-09 at 11 10 26 AM
> 
> Also, an idea I had is adding a target to our `Makefile` that can be invoked with `make create_key` that'll be an easy entrypoint for the CLI call with some defaults in place. Not a must, but it'd be an easy way for us to pull this branch down and test out the code locally.

Thanks for the commend @Olshansk! I just changed to the branch to `issue/utility_local_proof_of_stake_cli_RPC`. But I rebased to `deblasis:issue/utility_local_proof_of_stake_cli` before so there are still some delta. Do you want me to rebase it back to the "pocket/" branch?

For the `Makefile` do you want the new target be in a different PR?

**Comment by Olshansk** [comment]
> For the Makefile do you want the new target be in a different PR?

Same PR since its related. It's the tool we can use to verify your chang.

> Thanks for the commend @Olshansk! I just changed to the branch to issue/utility_local_proof_of_stake_cli_RPC. But I rebased to deblasis:issue/utility_local_proof_of_stake_cli before so there are still some delta. Do you want me to rebase it back to the "pocket/" branch?

The goal is really just to make sure that it's easy for the reviewer to see the changes, regardless of how we get there. In this specific case. I believe that @deblasis will merge #169 soon, so:

1. That gets merged in
2. Merge this with maser
3. Resolve conflicts
4. Push up changes
5. Make the base of this PR main
6. Let us, the reviewers, know the status of this PR or where you might need preliminary feedback (sync or async)

**Comment by Jasonyou1995** [comment]
> Any chance of minimizing this PR? I had expected the package and integration to be separate issues

Thank you for clarifying @andrewnguyen22 ! I think that's a good idea to separate so we can have a clear view to the changes.

I would like to get some of your feedbacks based on the current functions for Create, Delete, and Mnemonic. I will make some change accordingly.

**Comment by Jasonyou1995** [comment]
### Makefile `create_key` test | Delete key output
Screen Shot 2022-11-10 at 5 40 43 PM

### Delete none-exist key
Screen Shot 2022-11-10 at 5 41 02 PM

### Create a new key with given name
Screen Shot 2022-11-10 at 5 41 28 PM


Please let me know if there is change needed for the output format and features. There are some TODO listed in the source so we can discuss which ones need to be updated.

**PS: my updates are all in `share/keys/cmd**

File: shared/keys/cmd/create.go
Comment (andrewnguyen22) [inline shared/keys/cmd/create.go]: `TODO: (<Issue-#>): Import CryptoPocket`

Please add follow up issue numbers - also, why can't this be done `IN_THIS_COMMIT`?

File: shared/keys/cmd/delete.go
Comment (andrewnguyen22) [inline shared/keys/cmd/delete.go:92]: Remove comments or add comment for future use

File: shared/keys/cmd/root.go
Comment (andrewnguyen22) [inline shared/keys/cmd/root.go:62]: Please remove example comments and provide source

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]

File: shared/keys/cmd/create.go
Comment (Jasonyou1995) [reply shared/keys/cmd/create.go:1]: Thanks pointing out! I was not so sure about this neither. That license and copyright sections were generated by Cobra automatically.

File: shared/keys/LICENSE
Comment (Jasonyou1995) [reply shared/keys/LICENSE:1]: Yes, that is correct! I rebased to @deblasis 's latest director, and that's why it shows all the changes.

File: shared/keys/cmd/create.go
Comment (Jasonyou1995) [reply shared/keys/cmd/create.go]: Just updated and removed the unnecessary comments with future TODO

File: shared/keys/cmd/create.go
Comment (Jasonyou1995) [reply shared/keys/cmd/create.go]: Added error handling.

File: shared/keys/cmd/create.go
Comment (Jasonyou1995) [reply shared/keys/cmd/create.go]: Error handled

File: shared/keys/cmd/create.go
Comment (Jasonyou1995) [reply shared/keys/cmd/create.go:95]: Error handled

File: shared/keys/cmd/create.go
Comment (Jasonyou1995) [reply shared/keys/cmd/create.go]: I just have the crypto functions updated, and this is done in this commit now.

File: shared/keys/cmd/delete.go
Comment (Jasonyou1995) [reply shared/keys/cmd/delete.go:92]: Comments removed

**Comment by Jasonyou1995** [comment]
Just update codes based on review from @andrewnguyen22 . 

Currently, the module is fully isolated from the `pocket` module, so there is still some build error. Could you suggest how to update the module/package structure change?

I copied the `shared/crypto/` folder to my local directory, but I think that's redundant, so if you could suggest a way to update that would be great!

I used the functions `ED25519.go` to generate keys, so please check the validity of the key generating process in case anything is wrong.

**Comment by Olshansk** [comment]
> Currently, the module is fully isolated from the pocket module, so there is still some build error. Could you suggest how to update the module/package structure change?

Could you provide more details (screenshot? error) on what the build error is so we can point you in the right direction?

> I used the functions ED25519.go to generate keys, so please check the validity of the key generating process in case anything is wrong.

What do you mean by "validity" in this context?

> The mnemonic phrases is currently stored in the level.db along with other key information in JSON format. Please let me know if you want to remove any info from the key storage.

I'd personally store mnemonic phrases  in some dot file on the host machine, but I haven't looked into standard patterns. Does cosmos or do other projects use leveldb for it?

File: shared/keys/main.go
Comment (Olshansk) [inline shared/keys/main.go:1]: @Jasonyou1995 When we started working on #320, we ran into an issue where all methods in a directory and its subdirectories were different packages, because every directory in a Go package. Ref: https://www.codementor.io/@inanc/understanding-go-packages-cgsx1hzrx

This is a separate project (not a separate package), which needs a new main file, go.sum, go.mod, and there should only be one in any project. For example, VRF (https://github.com/pokt-network/pocket/tree/main/consensus/leader_election/vrf) is a package in the same project that is not integrated into the end-to-end flow.

Lmk if that makes sense. I can review the core work once this change is made.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: Makefile
Comment (Jasonyou1995) [reply Makefile:149]: I added it for the next PR so we don't forget.

File: .gitignore
Comment (Jasonyou1995) [reply .gitignore:33]: I think that is changed by @deblasis

File: Makefile
Comment (Jasonyou1995) [reply Makefile:153]: Yeah you are right, I'm updating the package structure so it can have all the dependencies live in a single go.mod file

File: shared/keys/main.go
Comment (Jasonyou1995) [reply shared/keys/main.go:1]: Yes! I will update the code for a better re-usability and merge all code from key-base into the pocket project. Thanks for pointing out and sharing the article!

**Review by Jasonyou1995 — COMMENTED** [review]
Reviewed items been updated or is updating

File: keys/cmd/delete.go
Comment (andrewnguyen22) [inline keys/cmd/delete.go:42]: Non-Blocker; Nit

Make inline (see example above)

File: keys/cmd/delete.go
Comment (andrewnguyen22) [inline keys/cmd/delete.go:56]: Non-Blocker; Nit

Make inline (see example above)

File: keys/cmd/delete.go
Comment (andrewnguyen22) [inline keys/cmd/delete.go:63]: `defer` means 'execute at the conclusion of the function

Thus, `defer kb.Close()` should be at line 40

File: keys/cmd/mnemonic.go
Comment (andrewnguyen22) [inline keys/cmd/mnemonic.go:29]: Handle error

File: keys/cmd/delete.go
Comment (andrewnguyen22) [inline keys/cmd/delete.go:22]: Try `ExactArgs` instead of `MinimumArgs`

File: keys/cmd/create.go
Comment (andrewnguyen22) [inline keys/cmd/create.go:139]: Ditto issue with `defer` (see below) watch out for this in other usages

File: keys/cmd/create.go
Comment (andrewnguyen22) [inline keys/cmd/create.go:54]: Remove unnecessary `err` declaration

File: keys/cmd/create.go
Comment (andrewnguyen22) [inline keys/cmd/create.go:22]: I'm not convinced the `CLI` portion of the keybase should be in this pacakge and not in the `app/client/cli` folder. WYT @Olshansk @deblasis

File: keys/cmd/create.go
Comment (andrewnguyen22) [inline keys/cmd/create.go:117]: It appears as if we are storing `clear text` `non obfuscated` private keys in a leveldb. This is not up to the V0 implementation's security requirements. 

[V0 Keybase Implementation](https://github.com/pokt-network/pocket-core/blob/40c52a1cf3d4ccc6ed161b2ac1853a003c32324f/crypto/keys/keybase.go#L238)

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]
A few nits, a few blockers. The biggest concern for me is the lack of security parity with V0. See comments below

**Review by Jasonyou1995 — COMMENTED** [review]
Updated comments from @Olshansk and @andrewnguyen22 . One blocker.

**Comment by deblasis** [comment]
@Jasonyou1995 regarding the `go.mod` and dependencies issues I have pulled your branch, made a few changes and now your logic is connected to the CLI. 

You can scan the delta of the changes here: 
https://github.com/pokt-network/pocket/compare/150-keybase-miniature-keybase-implementation-to-enable-m1-create-and-delete...150-keybase-miniature-keybase-implementation-to-enable-m1-create-and-delete_ADB_POC

**Do not merge it to your branch** 🙂 use it as an inspiration to understand what didn't work before and why. 

Bear in mind that I am simply addressing the cobra part and simply "made it work" in the fastest way possible to unlock you at least on the dependencies front, everything else still needs to be addressed. 

PTAL and revert with any questions you might have.

**Comment by Olshansk** [comment]
@Gustavobelfort No need to look at this right now, but just wanted to provide context that this work exists as a reference or starting point.

**Pull request closed without merging**
